airbyte-integrations/connectors/source-facebook-marketing/unit_tests/integration/test_ads_insights_action_product_id.py

```diff
@@ -0,0 +1,489 @@
+            "estimated_ad_recallers_upper_bound",
+            "frequency",
+            "full_view_impressions",
+            "full_view_reach",
+            "gender_targeting",
+            "impressions",
+            "inline_link_click_ctr",
+            "inline_link_clicks",
+            "inline_post_engagement",
+            "instant_experience_clicks_to_open",
+            "instant_experience_clicks_to_start",
+            "instant_experience_outbound_clicks",
+            "labels",
+            "location",
+            "mobile_app_purchase_roas",
+            "objective",
+            "optimization_goal",
+            "outbound_clicks",
+            "outbound_clicks_ctr",
+            "purchase_roas",
+            "qualifying_question_qualify_answer_rate",
+            "quality_ranking",
+            "reach",
+            "social_spend",
+            "spend",
+            "unique_actions",
+            "unique_clicks",
+            "unique_ctr",
+            "unique_inline_link_click_ctr",
+            "unique_inline_link_clicks",
+            "unique_link_clicks_ctr",
+            "unique_outbound_clicks",
+            "unique_outbound_clicks_ctr",
+            "updated_time",
+            "video_15_sec_watched_actions",
+            "video_30_sec_watched_actions",
+            "video_avg_time_watched_actions",
+            "video_continuous_2_sec_watched_actions",
+            "video_p100_watched_actions",
+            "video_p25_watched_actions",
+            "video_p50_watched_actions",
+            "video_p75_watched_actions",
+            "video_p95_watched_actions",
+            "video_play_actions",
+            "video_play_curve_actions",
+            "video_play_retention_0_to_15s_actions",
+            "video_play_retention_20_to_60s_actions",
+            "video_play_retention_graph_actions",
+            "video_time_watched_actions",
+            "website_ctr",
+            "website_purchase_roas",
+            "wish_bid",
+        ],
+        "time_increment": 1,
+        "action_attribution_windows": ["1d_click", "7d_click", "28d_click", "1d_view", "7d_view", "28d_view"],
+        "time_range": {"since": since, "until": until},
+    }
+    return RequestBuilder.get_insights_endpoint(access_token=ACCESS_TOKEN, account_id=ACCOUNT_ID).with_body(
+        encode_request_body(body)
+    )
+
+
+def _job_status_request(report_run_ids: Union[str, List[str]]) -> RequestBuilder:
+    if isinstance(report_run_ids, str):
+        report_run_ids = [report_run_ids]
+    body = {"batch": [{"method": "GET", "relative_url": f"{report_run_id}/"} for report_run_id in report_run_ids]}
+    return RequestBuilder.get_execute_batch_endpoint(access_token=ACCESS_TOKEN).with_body(encode_request_body(body))
+
+
+def _get_insights_request(job_id: str) -> RequestBuilder:
+    return RequestBuilder.get_insights_download_endpoint(access_token=ACCESS_TOKEN, job_id=job_id).with_limit(100)
+
+
+def _update_api_throttle_limit_response() -> HttpResponse:
```

Comment:
Do we test a case where the pct == 100? What happens in that case? Maybe this is nit and does not need to be tested now but it seems like a valid case and I wouldn't know how our source behaves in that case

Reply:
Good point, added a test for that, thanks!
Please see `test_given_api_throttle_exceeds_limit_on_first_check_when_read_then_wait_throttle_down_and_return_records`